connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultInboundConnectorContextFactory.java

```diff
@@ -0,0 +1,102 @@
+import java.lang.reflect.Type;
+import java.util.function.Consumer;
+
+public class DefaultInboundConnectorContextFactory implements InboundConnectorContextFactory {
+  private final ObjectMapper objectMapper;
+  private final InboundCorrelationHandler correlationHandler;
+  private final SecretProviderAggregator secretProviderAggregator;
+  private final ValidationProvider validationProvider;
+  private final OperateClientAdapter operateClientAdapter;
+  private final FeelEngineWrapper feelEngineWrapper;
+
+  public DefaultInboundConnectorContextFactory(
+      final ObjectMapper mapper,
+      final InboundCorrelationHandler correlationHandler,
+      final SecretProviderAggregator secretProviderAggregator,
+      final ValidationProvider validationProvider,
+      final OperateClientAdapter operateClientAdapter,
+      final FeelEngineWrapper feelEngineWrapper) {
+    this.objectMapper = mapper;
+    this.correlationHandler = correlationHandler;
+    this.secretProviderAggregator = secretProviderAggregator;
+    this.validationProvider = validationProvider;
+    this.operateClientAdapter = operateClientAdapter;
+    this.feelEngineWrapper = feelEngineWrapper;
+  }
+
+  @Override
+  public <T extends InboundConnectorExecutable<?>> InboundConnectorContext createContext(
+      final InboundConnectorDefinitionImpl newConnector,
+      final Consumer<Throwable> cancellationCallback,
+      final Class<T> executableClass) {
+
+    InboundConnectorContext inboundContext =
+        new InboundConnectorContextImpl(
+            secretProviderAggregator,
+            validationProvider,
+            newConnector,
+            correlationHandler,
+            cancellationCallback,
+            objectMapper);
+
+    if (isIntermediateContext(executableClass)) {
+      inboundContext =
+          new InboundIntermediateConnectorContextImpl(
+              inboundContext,
+              operateClientAdapter,
+              validationProvider,
+              feelEngineWrapper,
+              objectMapper);
+    }
+
+    return inboundContext;
+  }
+
+  private <T> boolean isIntermediateContext(Class<T> executableClass) {
+    // Retrieve all generic interfaces implemented by the executable class.
+    Type[] genericInterfaces = executableClass.getGenericInterfaces();
+
+    for (Type genericInterface : genericInterfaces) {
+      // Check if the current type is a parameterized type (i.e., has generic parameters).
+      if (genericInterface instanceof ParameterizedType parameterizedType) {
+        // Retrieve actual type arguments of the parameterized type.
+        Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
+
+        // If the first generic type argument matches InboundIntermediateConnectorContext,
+        // then this class represents an intermediate context.
+        if (actualTypeArguments.length > 0
+            && actualTypeArguments[0].equals(InboundIntermediateConnectorContext.class)) {
```

Comment:
To be on a safe side, suggesting the following:
```suggestion
            && InboundIntermediateConnectorContext.class.equals(actualTypeArguments[0])) {
```